File: cddlparser.py
import json
import sys
from cddlparser.parser import Parser
from cddlparser.ast import CDDLTree
from cddlparser.astencoder import ASTEncoder

def parse(string) -> CDDLTree:
    parser = Parser(string)
    return parser.parse()


if __name__ == "__main__":
    if len(sys.argv) <= 1:
        print("Please provide a CDDL file as parameter")
        print("Ex: python cddlparser.py tests/__fixtures__/example.cddl")
    else:
        file = sys.argv[1]
        with open(file, "r", encoding="utf8") as fhandle:
            cddl = fhandle.read()
            ast = parse(cddl)
            print("Abstract syntax tree (AST)")
            print("--------------------")
            print(ast)

            print()
            print("JSON serialization")
            print(json.dumps(ast, indent=2, cls=ASTEncoder))

            print()
            print("AST re-serialization")
            print("--------------------")
            print(ast.serialize())

File: typescript/src/cddlparser.ts
import path from "path";
import * as fs from "fs";
import { fileURLToPath } from "url";
import { Parser } from "./parser.ts";
import { CDDLTree } from "./ast.ts";

/**
 * Parse a CDDL string and return the resulting abstract syntax tree.
 *
 * @param string The CDDL string to parse
 * @returns The abstract syntax tree
 */
export function parse(string: string): CDDLTree {
  const parser = new Parser(string);
  return parser.parse();
}

/**
 * Main entry point for the script when run from the command-line.
 */
function main(): void {
  const args = process.argv.slice(2);
  if (args.length === 0) {
    console.log("Please provide a CDDL file as parameter");
    console.log("Ex: node typescript/cddlparser.js tests/__fixtures__/example.cddl");
  } else {
    const file = args[0]!;
    try {
      const cddl = fs.readFileSync(path.join('.', file), { encoding: "utf8" });
      const ast = parse(cddl);

      console.log("Abstract syntax tree (AST)");
      console.log("--------------------");
      console.log(ast.toString());

      console.log();
      console.log("JSON serialization");
      console.log(JSON.stringify(ast, null, 2));

      console.log();
      console.log("AST re-serialization");
      console.log("--------------------");
      console.log(ast.serialize());
    } catch (err) {
      console.error(`Error reading or parsing file: ${(err as Error).message}`);
    }
  }
}

// Run the main function if the script is executed directly
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  main();
}
